client: improve API of role and user listings

ark76r opened the pull request:
AuthRoleAPI.ListRoles() and AuthUserAPI.ListUsers() return values of type []string.
Because server returns json (eg.
{"roles":[{"role":"root","permissions":{"kv":{"read":["*"],"write":["*"]}}}]}),
the result cannot be unmarshaled into []string.

This patch changes these functions to return []Role and []User.

yichengq commented on etcdctl/command/role_commands.go:
```diff
@@ -101,7 +101,7 @@ func actionRoleList(c *cli.Context) {
 
 	for _, role := range roles {
-		fmt.Printf("%s\n", role)
+		fmt.Printf("%s\n", role.Role)
```
why do we need to change this?

ark76r replied:
Without it, etcdctl will print the dump of the whole structure what looks rather messy.
I think etcdctl should have additional command, eg. role show <role_name> to display details of the selected role.

I can prepare such code.

yichengq replied:
I would prefer to fix them in follow-up PRs to make this one focus on one thing.

yichengq commented:
The change on client API looks good to me. @barakmich Any thoughts about it?

yichengq commented on client/auth_user.go:
```diff
@@ -188,12 +188,13 @@ func (u *httpAuthUserAPI) ListUsers(ctx context.Context) ([]string, error) {
 	if err != nil {
 		return nil, err
 	}
+
```
remove this unnecessary new line.

yichengq commented:
You need to change the git commit title to follow instructions in https://github.com/coreos/etcd/blob/master/CONTRIBUTING.md#format-of-the-commit-message .

You have tested about the new API, right? does the unmarshal work?

ark76r commented:
@yichengq Yeah, I've tested it. I've created and deleted roles and users using etcdctl and tested with this:

    package client_test

    import (
        "testing"

        "github.com/coreos/etcd/client"
        "golang.org/x/net/context"
    )

    func TestListUsers(t *testing.T) {
        addr := []string{"http://127.0.0.1:2379"}
        config := client.Config{
            Endpoints: addr,
            Transport: client.DefaultTransport,
        }
        c, err := client.New(config)
        if err != nil {
            t.Errorf("Error creating client: %s", err)
        }
        api := client.NewAuthUserAPI(c)
        list, err := api.ListUsers(context.Background())
        for _, item := range list {
            if item.User == "" {
                t.Errorf("Something wrong with 'User'")
            }
        }
    }

    func TestListRoles(t *testing.T) {
        addr := []string{"http://127.0.0.1:2379"}
        config := client.Config{
            Endpoints: addr,
            Transport: client.DefaultTransport,
        }
        c, err := client.New(config)
        if err != nil {
            t.Errorf("Error creating client: %s", err)
        }
        api := client.NewAuthRoleAPI(c)
        list, err := api.ListRoles(context.Background())
        for _, item := range list {
            if item.Role == "" {
                t.Errorf("Something wrong with 'Role'")
            }
        }
    }

It works in all cases.

I'm going to follow your instructions and remove from this PR changes in etcdctl and set commit's title to
'client: fix results of role and user listings'

Technical question (cause I've never tried it on github): may I change the content of the commit using commit --amend?

yichengq commented:
@ark76r Could you also check that `list, err := api.ListUsers(context.Background())` return nil error?

I would prefer commit title to be something like `client: improve API of role and user listings`

> may I change the content of the commit using commit --amend?

Yes. After you change the commit, just force push it onto github.

ark76r commented:
@yichengq OK, fixed, title, changes only in the 'client' module. 

But, there is another (json umanrshalling error) problem and I think i's the problem on the server side: UserList after my changes returns value of the type []User.

This structure contain roles defined as (line 34 in auth_user.go)
 
     Roles    []string `json:"roles"`

But if user has granted at least one role, server (http://127.0.0.1:4001/v2/auth/users)  returns:

     {"users":[{"user":"my-user","roles":[{"role":"my-role","permissions":{"kv":{"read":["/other/path","/testpath"],"write":["/testpath/other"]}}}]}]}

And roles here are object (could be unmarshalled into the struct Role from auth_role.go).

In my opinion it is the error in server, which should return simple list of role names.

It cannot be resolved in the client module, because user of the AuthUserAPI probably expects that User object returned by UserList has the same type as User object returned by GetUser.

BTW, I don't know if GetUser works.

So what about this issue? Should be fixed on 'client' side?

yichengq commented:
@ark76r I think GetUser should return User struct with struct Role too. I check the auth API described [here](https://github.com/coreos/etcd/blob/master/Documentation/auth_api.md#users-1), and it does fit into Role struct.

Role is expanded so people could read it easily and it doesn't require another round of HTTP request.

ark76r commented:
@yichengq Documentation is inconsistent: starts from defining user as an object containing list of role names and immediately changes this definition returning user object containing list of full role definitions.

In my opinion this is error of the server API design.

Let's imagine simple scenario:
you have 3 roles, each describing privileges to 30 different paths.
Now you create 500 users, each has one to three of those roles.
API (as it works now) will return 500 times all the roles' definition.
Waste of memory, time, and network.

User and role definitions are separated not without a good reason.

Of course, in the client module it can be hidden by simple trick:
in the method UserList I can create another "user" structure (with roles defined as Roles []*Role) and then copy data to the User struct, stripping roles from useless data.

But in my opinion it is mistake in server API - it just shouldn't be done in this way.
In a real application I would read whole list of roles at the start, and then - querying users - I'm only interested in the information what roles are granted to the users.

For one user the present server API may look as an optimalization. For list of 1000 it's horrible bloat.

What is the decision? Futher changes in 'client' module, or my arguments look reasonable and changes should be done in server API?

raoofm commented:
@ark76r But as per your example, for a list of 1000 users, an admin trying to figure out which user has which role, it is a headache task, rather than getting all at 1 go. Though I understand your point, may be it could be handled with a flag.

yichengq commented:
@ark76r We decide to make it this way by referencing other existing APIs:
https://developer.github.com/v3/issues/#list-issues

Multiple HTTP requests may consume time for request sending and parsing.

> Of course, in the client module it can be hidden by simple trick:
in the method UserList I can create another "user" structure (with roles defined as Roles []*Role) and then copy data to the User struct, stripping roles from useless data.

I think a better way is to reuse the User struct. Any objections?

ark76r commented:
@yichengq OK, Definition of User structure is changes and now contains the field Roles []Role.

I've tested it with this code:

    func TestRolesAndUsers(t *testing.T) {
        addr := []string{"http://127.0.0.1:2379"}
        config := client.Config{
            Endpoints: addr,
            Transport: client.DefaultTransport,
        }
        c, err := client.New(config)
        if err != nil {
            t.Errorf("Error creating client: %s", err)
        }

        roleApi := client.NewAuthRoleAPI(c)
        _, err = roleApi.ListRoles(context.Background())
        if err != nil {
            t.Errorf("ListRoles error: %s", err)
        }

        const userName = "userOne"
        const userPassword = userName // :)

        userApi := client.NewAuthUserAPI(c)
        users, err := userApi.ListUsers(context.Background())
        if err != nil {
            t.Errorf("ListUsers error: %s", err)
        }
        for _, u := range users {
            if u.User == userName {
                userApi.RemoveUser(context.Background(), u.User)
            }
        }

        err = userApi.AddUser(context.Background(), userName, userPassword)
        if err != nil {
            t.Errorf("AddUser error: %s", err)
        }

        _, err = userApi.GetUser(context.Background(), userName)
        if err != nil {
            t.Errorf("GetUsers error: %s", err)
        }

    }

Looks like it works.
The next thing is the proper displaying it in etcdctl (and proper commands and flags).

yichengq commented:
The test code looks good to me except that you need to check the error returned by RemoveUser. BTW, i like your password. :wink:

ark76r commented:
OK, here you are - test source code is updated.

Of course with this API etcdctl will not compile:

    command\user_commands.go:195: cannot use newUser.Roles (type []client.Role) as type []string in argument to sort.Strings
    command\user_commands.go:196: cannot use currentUser.Roles (type []client.Role) as type []string in argument to sort.Strings
    command\user_commands.go:222: cannot use user.Roles (type []client.Role) as type []string in argument to strings.Join

etcdctl must be updated as well.

ark76r commented:
@raoofm Admin doesn't use the API. Presentation of the data must be defined and implemented in etcdctl tool.

yichengq commented:
@ark76r ok. we should update etcdctl to make it buildable.

raoofm commented:
@ark76r true, agree

ark76r commented:
@yichengq 
After all, I belive this is the best solution. (Of course I still don't like JSON returned by server in case of UserList command).

etcdctl works without changes (but need better data presentation).
This testcase suggests it works:

    package client_test

    import (
        "fmt"
        "testing"

        "golang.org/x/net/context"

        "github.com/coreos/etcd/client"
    )

    func TestUserRole(t *testing.T) {
        cfg := client.Config{
            Endpoints: []string{"http://127.0.0.1:4001"},
        }
        c, err := client.New(cfg)
        checkError("client.New", err, t)

        roleApi := client.NewAuthRoleAPI(c)
        userApi := client.NewAuthUserAPI(c)

        _ = roleApi.RemoveRole(context.Background(), "roleA") // cleanup
        _ = roleApi.RemoveRole(context.Background(), "roleB") // cleanup
        _ = userApi.RemoveUser(context.Background(), "userA") // cleanup
        _ = userApi.RemoveUser(context.Background(), "userB") // cleanup

        // tests

        err = roleApi.AddRole(context.Background(), "roleA")
        checkError("AddRole('roleA')", err, t)
        err = roleApi.AddRole(context.Background(), "roleB")
        checkError("AddRole('roleB')", err, t)

        paths := []string{"/roleA/path1/*", "/roleA/path2/*"}
        _, err = roleApi.GrantRoleKV(context.Background(), "roleA", paths, client.ReadPermission)
        checkError("GrantRoleKV", err, t)

        roles, err := roleApi.ListRoles(context.Background())
        checkError("ListRoles", err, t)
        var rolesDict map[string]client.Role = make(map[string]client.Role, len(roles))
        for _, rl := range roles {
            rolesDict[rl.Role] = rl
            fmt.Println("Role", rl.Role, "-", rl)
        }

        err = userApi.AddUser(context.Background(), "userA", "userA")
        checkError("addUser", err, t)

        err = userApi.AddUser(context.Background(), "userB", "userB")
        checkError("addUser", err, t)

        err = userApi.RemoveUser(context.Background(), "userB")
        checkError("RemoveUser", err, t)

        user, err := userApi.GetUser(context.Background(), "userB")
        if err == nil {
            t.Error("GetUser didn't return error 'not exist'")
        }
        if user != nil {
            t.Error("GetUser returned non-existent user", user)
        }

        user, err = userApi.GetUser(context.Background(), "userA")
        checkError("GetUser", err, t)

        user, err = userApi.GrantUser(context.Background(), "userA", []string{"roleA", "roleB"})
        checkError("GrantUser", err, t)

        user, err = userApi.RevokeUser(context.Background(), "userA", []string{"roleB"})
        checkError("RevokeUser", err, t)
        fmt.Println("user:", user)

        userList, err := userApi.ListUsers(context.Background())
        checkError("UserList", err, t)

        for _, usr := range userList {
            fmt.Println("User", usr.User, "-", usr)
            for _, rl := range usr.Roles {
                fmt.Println("- ", rl, rolesDict[rl])
            }
        }
    }

    func checkError(funcName string, err error, t *testing.T) {
        if err != nil {
            t.Error(funcName, err)
        }
    }

yichengq commented:
> After all, I belive this is the best solution.

I tried to reuse the existing User struct, but failed when calling GrantUser and RevokeUser. Is this the reason that you use a in-function User struct to unmarshal?

ark76r commented:
@yichengq Yes. Grant/RevokeUser expect Roles to be of type []string and they try to sort them using strings.Sort.

Of course I can change the implementation of the Grant/RevokeUser, but anyway - somewhere []Role must be just converted into []string.

BTW: look at my recent test code. You really don't need role definition in the user object.

yichengq commented on etcdctl/command/user_commands.go:
```diff
@@ -181,20 +181,28 @@ func userGrantRevoke(c *cli.Context, grant bool) {
 	api, user := mustUserAPIAndName(c)
 	currentUser, err := api.GetUser(ctx, user)
-	if currentUser == nil {
+	if currentUser == nil || err != nil {
```
just `err != nil`

yichengq commented:
I think this is the best approach considering current situation. I would like to merge it and move quickly.

LGTM except the comment.

You have tested it using your test code, right?

ark76r commented:
@yichengq I've made this single small change in the user_commands.go file following your comment
(now it its: if err!= nil).

Yes, I've tested it. Here you have the fill sourcecode of my tests: https://gist.github.com/ark76r/dfdf766b4425617ee2ba

Of course, I've also used etcdctl from command line. Now it is hardly usable (prints data in horrible format) but there is no json unmarshal error, and after all it's also looks like in normal situation there is no panics of things like that.

So I expect there will be changes in server API to make it more consisten. In about 12h I will try to suggest changes (of course it's only suggestion and you can do with it whatever you want)

The other thing is that server has some problems with the semantics of some commands.

Look at ma tests sourcecode. The second test (TestServerBehaviour ) gives such results:

    --- FAIL: TestServerBehaviour (0.30s)
	arprv_test.go:112: >>> Add nonexistent user
	arprv_test.go:122: >>> Try add the user, who already exists
	arprv_test.go:125: AddUser of existing user didn't return the error. What is the semantics of this fn?
	arprv_test.go:128: >>> Grant nonexistent role to the user
	arprv_test.go:131: Added nonexistent role to the user successed. What does it mean?
	arprv_test.go:139: User has some possibly nonexistent roles:
	arprv_test.go:141: thereisnosuchrole
	arprv_test.go:147: >>> GetUser on the user who has nonexistent role
	arprv_test.go:233: GetUser error: auth: Role thereisnosuchrole does not exist.
	arprv_test.go:165: >>> Get the list of the users, and one of the has nonexistent role
	arprv_test.go:168: - ListUser has returned error: auth: Role thereisnosuchrole does not exist.
	arprv_test.go:179: >>> Revoking unexistent role from the user
	arprv_test.go:183:   - OK
	arprv_test.go:197: Try to revoke a role from the user he has never had
	arprv_test.go:202: Revoking the role, the user didn't have is an error:  auth: Revoking ungranted role nevergotit for user userB
	arprv_test.go:205: >>> Create new role: roleC
	arprv_test.go:209: >>> Grant roleC to userB
	arprv_test.go:213: >>> Remove role granted previously to the user
	arprv_test.go:216: - RemoveRole didn't returned the error. Let's check GetUser in such case
	arprv_test.go:219: After removing role granted to the user, GetUser returned error: auth: Role roleC does not exist.

 In my opinion, the problems are:

1. AddUser('existentUser') - the result doesn't return the error. No info if I've added user or not. I don't know what about the password (maybe it changes the password mayby not).
2. GrantUser('user', 'nonexistent role') - there is no error returned, but than operations like GetUser ListUser return errors.
3. RemoveRole('roleGrantedToUsers') - role is removed and users become invalid (subsequent calls to GetUser ListUser return errors).

The funny thing is that you cannot revoke from the user the nonexistent role from etcdctl tool but you can do it from API.
etcdctl calls GetUser and get error here so it desn't calls UserRevoke.
In API I can skip the call to GetUser and call first off UserRevoke, so I can revoke that nonexistent role from the user.

In general, I belive the model of the users and roles in the server isn't completed and consistent and requires rethinking.

I'm going to prepare the paper with some ideas today (later). Maybe it will be worth of discussion.

xiang90 commented:
I believe this is already fixed by @heyitsanthony's pr.
